lib/matplotlib/axis.py

```diff
@@ -1498,6 +1498,10 @@ def have_units(self):
         return self.converter is not None or self.units is not None
 
     def convert_units(self, x):
+        # If x is already a number, doesn't need converting
+        if munits.ConversionInterface.is_numlike(x):
```

Comment:
Right now the assigned converter deals with the floats. 

This change means that a user can *never* create a converter that accepts `numlike` vectors and does something with them.  They'd always have to wrap it in a class that `is_numlike` doesn't think is numlike.  

I guess I'm OK w/ that restriction, and I don't think it harms any of our current converters.  But, I'm not 100% sure what the rest of the world is doing.

Reply:
BTW pinging @tacaswell, @anntzer and @story645  as they are working on categoricals where unit handling has come up a lot

Reply:
Hmm, you make a good point. **If** the decision is to allow plotting numbers on an axis with units, then this PR is the way to go. (In principle I'm -1 on that, but I wasn't on the call so I'm guessing there's good reasons for allowing it?)

Reply:
I think there are good reasons for both. 

@efiring 's example was: imagine you plot five categories, and then want a line at x=0.5.  Thats hard to do if we have locked out x to be only categories.

I'm not 100% convinced we can't come up w/ a way to let the user *force* the units, but still retain the lockout by default.

Reply:
(I wasn't really "working on categoricals", just trying to write some version that's reasonably robust and doesn't do too silly things (which is quite hard) :-))

Reply:
Another example: you might have some variables (e.g., big arrays) already represented as mpl datenums, and also want to be able to position text, or some symbol, using a Datetime specification. The axis is in datetime mode.  Why should you have to either convert your big array back to the incredibly clunky array of Datetime objects, or manually convert your text or symbol positions to datenums?

Reply:
Wasn't this supposed to be the difference between methods like axvline and blinds? Axis space vs coordinate space. I think this runs into all the same complaints of supporting mixed types in categorical, namely if I plot a categorical ['1', '2', '3', '4', '5'], that 5 will map to integer 4 and so not line up with an integer 5 plot anyway.

Reply:
But `axvline('5')` would work and line up with the 5 category right?

Reply:
Was there a decision on supporting mixed types in categorical?

Reply:
Just strings, according to #10212

Reply:
Axvline plots a line at an x in data coordinates.  If you want that half-way between two categorical values, you have to be able to specify that location in *numeric* data coordinates.  If you know how categoricals are mapped to numbers, and you are *allowed* to plot in numeric data coordinates, this is easy.  Fundamentally, all plotting is done in floating point coordinates.  Plotting something in "units" should not prevent the user from positioning something directly in the numbers to which those units map--and from doing so efficiently.

Reply:
Regarding your example, it should be well-documented--if it isn't already--that categoricals map sequentially to integers starting from zero.  Knowing that, the user can easily calculate that the point half-way between your '2' and '3' (which more realistically might be 'dogs' and 'cats' or "instrument 872" and "instrument 253") is 1.5.